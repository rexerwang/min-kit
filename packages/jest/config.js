/**
 * @typedef {import('jest').Config Config}
 */

const defineJestConfig = require('./config/taro-jest/index').default

/** @type {Config} */
const baseConfig = defineJestConfig({
  globalSetup: require.resolve('./config/globalSetup.js'),
  setupFilesAfterEnv: [require.resolve('./config/setupAfterEnv.js')],
  testEnvironment: '@happy-dom/jest-environment', // 'jsdom'
  testMatch: ['<rootDir>/src/__tests__/**/*.test.ts?(x)'],
  collectCoverageFrom: ['<rootDir>/src/**/*.ts?(x)', '!**/__tests__/**'],
  coverageReporters: process.env.CI ? ['json'] : ['html', 'text'],
  passWithNoTests: true,
})

/**
 * defineJestConfig
 * @param {Partial<Config>} [config]
 * @returns {Config}
 */
module.exports = (config = {}) => {
  let { setupFilesAfterEnv, collectCoverageFrom, ...configs } = baseConfig

  if (Array.isArray(config.setupFilesAfterEnv)) {
    setupFilesAfterEnv = setupFilesAfterEnv.concat(config.setupFilesAfterEnv)
  }

  if (Array.isArray(config.collectCoverageFrom)) {
    collectCoverageFrom = [collectCoverageFrom[0], ...config.collectCoverageFrom, collectCoverageFrom[1]]
  }

  return {
    ...configs,
    ...config,
    setupFilesAfterEnv,
    collectCoverageFrom,
  }
}
