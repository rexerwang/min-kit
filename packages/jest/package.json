{
  "name": "@min-kit/jest",
  "version": "0.5.11",
  "private": true,
  "description": "jest config & utils for internal",
  "author": "rexerwang",
  "license": "MIT",
  "repository": "rexerwang/min-kit",
  "exports": {
    ".": "./utils.ts",
    "./config": "./config.js"
  },
  "devDependencies": {
    "@happy-dom/jest-environment": "^14.7.1",
    "@tarojs/test-utils-react": "^0.1.1",
    "@testing-library/jest-dom": "^6.4.8",
    "@testing-library/react": "^16.0.1"
  }
}
