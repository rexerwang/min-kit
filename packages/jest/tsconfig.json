{
  "extends": "../../tsconfig.root.json",
  "include": ["utils.ts"]
}
