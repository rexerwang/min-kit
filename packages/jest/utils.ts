import TestUtils from '@tarojs/test-utils-react'

export const createTaroTestUtils = () => {
  const testUtils = new TestUtils()
  return Object.assign(testUtils, { unmount: testUtils.unmout })
}

export const spyOnConsole = (methods = ['error', 'warn', 'info', 'debug']) =>
  methods.map((method: any) => {
    const spy = jest.spyOn(console, method)
    spy.mockImplementation(jest.fn())
    return spy
  })

export * from '@testing-library/react'
