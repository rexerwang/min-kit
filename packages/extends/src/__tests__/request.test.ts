import { spyOnConsole } from '@min-kit/jest'
import Taro from '@tarojs/taro'

const baseUrl = 'https://request-api.com/api'

describe('request', () => {
  const requestSpy = jest.spyOn(Taro, 'request')
  const abortStub = jest.fn()

  beforeEach(() => {
    requestSpy.mockImplementation((request) => {
      const response = Object.assign(
        Promise.resolve({
          statusCode: Number(request.url.split('?').at(0)!.split('/').at(-1)) || 200,
          data: request.data ?? {},
          header: {},
          errMsg: 'ok',
        }),
        { abort: abortStub },
      )
      return response as any
    })
  })

  afterEach(() => {
    requestSpy.mockClear()
    abortStub.mockClear()
  })

  it.each(['get', 'delete', 'head', 'options', 'post', 'put', 'patch'])(
    'should request.%s successfully when given statusCode = 200',
    async (method) => {
      const { createRequest } = await import('../request')
      const http = createRequest({ baseUrl })

      const res = await http[method]('/200', { payload: 'test' })
      expect(res.statusCode).toBe(200)
      expect(requestSpy).toHaveBeenCalledWith({
        url: baseUrl + '/200',
        method: method.toUpperCase(),
        data: {
          payload: 'test',
        },
        header: {},
      })
    },
  )

  it.each(['get', 'delete', 'head', 'options', 'post', 'put', 'patch'])(
    'should request.%s failed when given statusCode = 500',
    async (method) => {
      spyOnConsole('error')

      const { createRequest } = await import('../request')
      const http = createRequest({ baseUrl })

      await expect(() => http[method]('/500', { payload: 'test' })).rejects.toThrow(
        `${method.toUpperCase()} ${baseUrl}/500 500 (ok)`,
      )
    },
  )

  it('should execute middlewares successfully', async () => {
    const { createRequest } = await import('../request')
    const http = createRequest({ baseUrl })

    http.use(async (ctx, next) => {
      ctx.set('x-test', 'test')
      ctx.set('x-abc')
      ctx.request.data = { ...ctx.request.data, a: 'b' }

      await next()
    })

    await http.post('/200', { payload: 'test' }, { header: { 'x-abc': 'abc' } })
    expect(requestSpy).toHaveBeenCalledWith({
      url: baseUrl + '/200',
      method: 'POST',
      data: {
        payload: 'test',
        a: 'b',
      },
      header: {
        'x-test': 'test',
      },
    })
  })

  it('should replay request correctly in middleware', async () => {
    const { createRequest } = await import('../request')
    const http = createRequest({ baseUrl })

    http.use(async (ctx, next) => {
      await next()

      if (!ctx.request.replayed) {
        await ctx.replay()
      }
    })

    const res = await http.get('/200')
    expect(requestSpy).toHaveBeenCalledTimes(res.request.replayed! + 1)
  })

  it('should abort request did work in middleware', async () => {
    const { createRequest, AbortControllerImpl } = await import('../request')

    const http = createRequest({ baseUrl })
    const abort = new AbortControllerImpl()

    await Promise.all([
      http.get('/200', undefined, { abort }),
      Promise.resolve().then(() => abort.abort()),
      Promise.resolve().then(() => abort.abort()), // abort duplicated
    ])

    expect(abortStub).toHaveBeenCalledTimes(1)
  })

  it.each([
    'call ctx.abort() before next()',
    'call ctx.replay() before next()',
    'call next() duplicated',
    'call ctx.set() after next()',
    'call ctx.abort() after next()',
  ])('should throw error when %p in middleware', async (type) => {
    const { createRequest } = await import('../request')
    const http = createRequest({ baseUrl })

    http.use(async (ctx, next) => {
      if (type.includes('before')) {
        if (type.includes('abort')) ctx.abort()
        else if (type.includes('replay')) await ctx.replay()
      }

      await next()

      if (type.includes('duplicated')) await next()
      else if (type.includes('after')) {
        if (type.includes('set')) ctx.set('x-abc', 'abc')
        else if (type.includes('abort')) ctx.abort()
      }
    })

    await expect(http.post('/200')).rejects.toThrowErrorMatchingSnapshot()
  })
})
