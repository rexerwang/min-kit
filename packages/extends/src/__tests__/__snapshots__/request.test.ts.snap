// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`request should throw error when given invalid middleware usage: "call abort() after next()" 1`] = `"Cannot abort request after next()."`;

exports[`request should throw error when given invalid middleware usage: "call abort() before next()" 1`] = `"request aborted"`;

exports[`request should throw error when given invalid middleware usage: "call next() duplicated" 1`] = `"next() called multiple times"`;

exports[`request should throw error when given invalid middleware usage: "call replay() before next()" 1`] = `"Cannot replay request before next()."`;

exports[`request should throw error when given invalid middleware usage: "call set() after next()" 1`] = `"Cannot set request headers after next()."`;
